source/azure_iot_provisioning_client.c

```diff
@@ -0,0 +1,963 @@
+    az_span xTopic = az_span_create( xAzureIoTProvisioningClientHandle->_internal.pucProvisioningLastResponseTopic,
+                                     ( int32_t ) xAzureIoTProvisioningClientHandle->_internal.usProvisioningLastResponseTopicLength );
+
+    /* Check the state.  */
+    if( xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState == azureiotprovisioningWF_STATE_RESPONSE )
+    {
+        xCoreResult =
+            az_iot_provisioning_client_parse_received_topic_and_payload( &xAzureIoTProvisioningClientHandle->_internal.xProvisioningClientCore,
+                                                                         xTopic, xPayload, &xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse );
+
+        if( az_result_failed( xCoreResult ) )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningFailed );
+            AZLogError( ( "IoTProvisioning client failed to parse packet, error status: %08x", xCoreResult ) );
+            return;
+        }
+
+        if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.operation_status == AZ_IOT_PROVISIONING_STATUS_ASSIGNED )
+        {
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningSuccess );
+        }
+        else if( xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.retry_after_seconds == 0 )
+        {
+            AZLogError( ( "IoTProvisioning client failed with error %d, extended error status: %u\r\n",
+                          xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.error_code,
+                          xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.registration_state.extended_error_code ) );
+            /* Server responded with error with no retry.  */
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningServerError );
+        }
+        else
+        {
+            xAzureIoTProvisioningClientHandle->_internal.ullProvisioningRequestTimeout =
+                xAzureIoTProvisioningClientHandle->_internal.xGetTimeFunction() + xAzureIoTProvisioningClientHandle->_internal.xRegisterResponse.retry_after_seconds;
+            prvAzureIoTProvisioningClientUpdateState( xAzureIoTProvisioningClientHandle, eAzureIoTProvisioningPending );
+        }
+    }
+    else
+    {
+        AZLogWarn( ( "Parse response action called in wrong state %d\r\n",
+                     xAzureIoTProvisioningClientHandle->_internal.ulWorkflowState ) );
+    }
+}
+/*-----------------------------------------------------------*/
+
+/**
+ *
+ * Implementation of wait check action, this action is only allowed in azureiotprovisioningWF_STATE_WAITING
+ *
+ * */
+static void prvAzureIoTProvisioningClientCheckTimeout( AzureIoTProvisioningClientHandle_t xAzureIoTProvisioningClientHandle )
```

Comment:
1 -- 
What happens if (a) our provisioning request "fails" but the server sends a "try again in 20 minutes" or something kind of long?  It naively seems to me that we will stay in the WAITING state of the state machine - which seems reasonable, except I would worry that we'll never invoke the underlying MQTT layer to do an MQTT PING and the connection could end up being dropped on us.

Is this a reasonable read / concern?

2 -- Should we have a mechanism to return `ullProvisioningRequestTimeout` back to the application?  For instance if Hub says "I'll be ready in 10 hours" that could impact a sophisticated application's desire to retry.  Though I don't know if DPS would return such a long delay in 1st place or error out?

FWIW there's no harm in NOT having a `ullProvisioningRequestTimeout` returner, since we can always add later, but want to at least see your thinking on it?

Reply:
Currently with each action trigger, we also call ProcessLoop of MQTT, so that should take care of PingReq/PingRes.

For 2, I would suggest to add it latter.


---
In reply to: [603546950](https://github.com/Azure/azure-iot-middleware-freertos/pull/49#discussion_r603546950) [](ancestors = 603546950)